core/src/main/java/hudson/Util.java

```diff
@@ -304,19 +337,144 @@ private static void makeWritable(@Nonnull File f) {
 
     }
 
+    /**
+     * Deletes the given directory (including its contents) recursively.
+     * It does not take no for an answer - if necessary, it will have multiple
+     * attempts at deleting things.
+     *
+     * @throws IOException
+     * if the operation fails.
+     */
     public static void deleteRecursive(@Nonnull File dir) throws IOException {
+        for( int numberOfAttempts=1 ; ; numberOfAttempts++ ) {
+            try {
+                tryOnceDeleteRecursive(dir);
+                break; // success
+            } catch (IOException ex) {
+                boolean threadWasInterrupted = pauseBetweenDeletes(numberOfAttempts);
+                if( numberOfAttempts>= DELETION_MAX || threadWasInterrupted)
+                    throw new IOException(deleteFailExceptionMessage(dir, numberOfAttempts, threadWasInterrupted), ex);
+            }
```

Comment:
Minor: The `deleteFile`, `deleteRecursive` and `deleteContentsRecursive` all look like they share similar method bodies. Is it possible to merge them all to reduce copy-paste code?